for_testing.py

```diff
@@ -0,0 +1,130 @@
+#creating an empty layer
+cLayer = qgis.utils.iface.mapCanvas().currentLayer()
+provider = cLayer.dataProvider()
+#print (provider.encoding(),",", provider.fields(),",", provider.crs())
+
+outputpath = "D:/M.Tech/Fifth Sem/MTP/snr/sinnar_dg.shp"
+
+# SaveVectorOptions contains many settings for the writer process
+# Write to an ESRI Shapefile format dataset using UTF-8 text encoding
+save_options = QgsVectorFileWriter.SaveVectorOptions()
+save_options.driverName = "ESRI Shapefile"
+save_options.fileEncoding = "UTF-8"
+transform_context = QgsProject.instance().transformContext()
+writer = QgsVectorFileWriter.writeAsVectorFormatV3(cLayer,
+                                                 outputpath,
+                                                  transform_context,
+                                                  save_options) #[Not sure if this one is right or wrong]
+
+layer = QgsVectorLayer(outputpath,"rs02082022", "ogr")
+
+if not layer.isValid():
+    print ("Layer failed to load!")
+
+#start editing
+layer.startEditing()
+provider = layer.dataProvider()
+
+#add fields
+provider.addAttributes( [ QgsField("AddressFrom",  QVariant.String)] )
+provider.addAttributes( [ QgsField("AddressTo",  QVariant.String)] )
+layer.commitChanges()
+
+# prepare graph
+vectorLayer  = QgsVectorLayer ("/D:/M.Tech/Fifth Sem/MTP/snr/Transport_Road.shp",'lines')
+director = QgsVectorLayerDirector( vectorLayer, -1, '', '', '', 3 )
+strategy = QgsNetworkDistanceStrategy()
+director.addStrategy( strategy )
+
+builder = QgsGraphBuilder(vectorLayer.crs()) #[ERROR]
+
+# [From this point the code is not working properly.]
+# prepare points
+features = processing.features(point_layer) #[ERROR]
+point_count = point_layer.featureCount() #It's showing 0 points means 
+                                        #I did something wrong while debugging but what was that I am not sure.
+
+points = []
+
+for f in features:
+  points.append(f.geometry().asPoint())
+
+tiedPoints = director.makeGraph( builder, points )
+graph = builder.graph()
+
+for i in range(0,point_count-1):
+    for j in range (i+1,point_count ):
+            
+    #for i in range(0,5):    
+    #    progress.setPercentage(int(100 * i/ point_count))
+        
+        from_point = points[i]
+        to_point = points[j]
+    
+        from_id = graph.findVertex(from_point)
+        to_id = graph.findVertex(to_point)
+    
+        (tree,cost) = QgsGraphAnalyzer.dijkstra(graph,from_id,0)
+    
+        if tree[to_id] == -1:
+            continue # ignore this point pair
+        else:
+            # collect all the vertices between the points
+            route_points = []
+            curPos = to_id 
+            while (curPos != from_id):
+                route_points.append( graph.vertex( graph.arc( tree[ curPos ] ).inVertex() ).point() )
+                curPos = graph.arc( tree[ curPos ] ).outVertex()
+     
+            route_points.append(from_point)

```

Comment:
```
            while (curPos != from_id):
                route_points.append( graph.vertex( graph.edge(tree[curPos]).fromVertex().point() ) )
                curPos = graph.edge( tree[ curPos ] ).outVertex()
                #print(curPos)
            route_points.append(QgsPoint(points[i]))
            # route_points.append(from_point)
            #print(route_points)
```
       `route_points.append( graph.vertex( graph.edge(tree[curPos]).fromVertex().point() ) )`

This thing returns me list index out of range because the `to_id` passing -1.

I was trying to solve this but can not able to do that.